### PR TITLE
[Merged by Bors] - feat: rewrite updating style-exceptions in Lean

### PR DESCRIPTION
**Pull request by grunweg**
Rewrite the `update-style-exceptions` script in Lean, by adding an --update flag to `lint_style`:
in this mode, the script will remove any exceptions which are no longer needed,
add new entries for any current error, and try to leave existing entries untouched.
(With one exception: if a "file too long" error still applies, but the file is currently at least 200 lines shorter,
we update it to the new entry.)

---

Commits can mostly be reviewed individually: the second last commit fixes a bug in a previous commit; as the intermediate commits include a merge, I haven't tried to combine these. You can try to find the bug if you like, or review the whole diff for that part.
Second iteration of #14273, i.e. the third iteration of this in total.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
bors merge

**Review by adomani — COMMENTED** [review]
Just two very small comments!

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 1]
```diff
@@ -109,22 +109,52 @@ structure ErrorContext where
...
+  /-- The existing exception also covers the new error.
+  Indicate whether we prefer keeping the existing exception (the more common case)
+  or would rather replace it by the new exception
+  (this is more rare, and currently only happens for particular file length errors). -/
+  | Comparable (preferExisting : Bool)
+  deriving BEq
+
+/-- Determine whether a `new` `ErrorContext` is covered by an `existing` exception,
+and, if it is, if we prefer replacing the new exception or keeping the previous one. -/
+def compare (existing new : ErrorContext) : ComparisonResult := Id.run do
```
[the remark this answers is not in the document]
Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean:128]: Nice catch! You are right; I removed this.

**Comment by PatrickMassot** [comment]
bors merge

**Pull request closed without merging**